### PR TITLE
Added defensive code for scenario where thread id <= 0

### PR DESCRIPTION
**Pull request by dhoard**
Signed-off-by: Doug Hoard <doug.hoard@gmail.com>

### CODE REVIEW

**Comment by dhoard** [comment]
PR to resolve https://github.com/prometheus/client_java/pull/784 which will ultimately resolves https://github.com/prometheus/jmx_exporter/issues/601

@fstab please review

**Comment by dhoard** [comment]
@fstab 

**TL;DR**

ZooKeeper class `QuorumPeer` derives from `Thread` and overrides `getId()` based on the ZooKeeper node id (myid = `0`). (I don't have the history as to why ZooKeeper chose to override the thread id... but it's been in their code code for at least 4+ years.)

I feel that it shouldn't break `ThreadExports`.

**Detailed Analysis**

https://github.com/prometheus/jmx_exporter/issues/601

**ThreadExports Test Case**

```
  @Test
  public void testThreadId0() {
    Runnable runnable = new Runnable() {
      @Override
      public void run() {
        try {
          Thread.sleep(10000);
        } catch (InterruptedException e) {
          // DO NOTHING
        }
      }
    };

    Thread thread = new MyThread(runnable);
    thread.setName("test");
    thread.start();

    ThreadExports threadExports = new ThreadExports();
    threadExports.collect();
  }

  class MyThread extends Thread {

    public MyThread(Runnable runnable) {
      super(runnable);
    }

    public long getId() {
      return 0;
    }
  }
```

**Test Case Result (without the fix)**

```
testThreadId0(io.prometheus.client.hotspot.ThreadExportsTest)  Time elapsed: 0.142 sec  <<< ERROR!
java.lang.IllegalArgumentException: Invalid thread ID parameter: 0
	at java.management/sun.management.ThreadImpl.verifyThreadId(ThreadImpl.java:165)
	at java.management/sun.management.ThreadImpl.verifyThreadIds(ThreadImpl.java:174)
	at java.management/sun.management.ThreadImpl.getThreadInfo(ThreadImpl.java:180)
	at io.prometheus.client.hotspot.ThreadExports.getThreadStateCountMap(ThreadExports.java:146)
	at io.prometheus.client.hotspot.ThreadExports.addThreadMetrics(ThreadExports.java:113)
	at io.prometheus.client.hotspot.ThreadExports.collect(ThreadExports.java:177)
	at io.prometheus.client.hotspot.ThreadExports.collect(ThreadExports.java:171)
	at io.prometheus.client.hotspot.ThreadExportsTest.testThreadId0(ThreadExportsTest.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
```

**Review by fstab — CHANGES_REQUESTED** [review]
Thanks a lot for the explanation. That's unexpected. I agree we should handle this and not just fail in that case. I left a few review comments. Thanks for picking this up!

File: simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/ThreadExports.java [hunk 1]
```diff
@@ -121,8 +122,26 @@ void addThreadMetrics(List<MetricFamilySamples> sampleFamilies, Predicate<String
...
+    int index = 0;
+    threadIds = new long[threadIdSet.size()];
+    for (Long threadId : threadIdSet) {
+      threadIds[index] = threadId;
+      index++;
+    }
+
     // Get thread information without computing any stack traces
-    ThreadInfo[] allThreads = threadBean.getThreadInfo(threadBean.getAllThreadIds(), 0);
+    ThreadInfo[] allThreads = threadBean.getThreadInfo(threadIds, 0);
```
[the remark this answers is not in the document]
Comment (dhoard) [reply simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/ThreadExports.java]: Agreed. Will refactor.

File: simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java [hunk 2]
```diff
@@ -83,49 +88,63 @@ public HTTPMetricHandler(CollectorRegistry registry, Supplier<Predicate<String>>
...
-            if ("/-/healthy".equals(contextPath)) {
-                osw.write(HEALTHY_RESPONSE);
-            } else {
-                String contentType = TextFormat.chooseContentType(t.getRequestHeaders().getFirst("Accept"));
-                t.getResponseHeaders().set("Content-Type", contentType);
-                Predicate<String> filter = sampleNameFilterSupplier == null ? null : sampleNameFilterSupplier.get();
-                filter = SampleNameFilter.restrictToNamesEqualTo(filter, parseQuery(query));
-                if (filter == null) {
-                    TextFormat.writeFormat(contentType, osw, registry.metricFamilySamples());
+        public void handle(HttpExchange httpExchange) {
```
[the remark this answers is not in the document]
Comment (dhoard) [reply simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java]: Removed the code change. I think we should at least log the issue... but that can be tabled for another issue.

File: simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/ThreadExports.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhoard) [reply simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/ThreadExports.java]: Agreed. Made the code change.

**Comment by fstab** [comment]
Thank you!

**Pull request merged**